### PR TITLE
Contributors util

### PR DESCRIPTION
**Pull request by phlax**
Add a contributors util to retrieve per-user contributor counts optionally filtered by time

### CODE REVIEW

File: pootle/apps/pootle_statistics/utils.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+from django.contrib.auth import get_user_model
+from django.utils.functional import cached_property
+
+
+User = get_user_model()
+
+
+class Contributors(object):
+
+    def __init__(self, include_anon=False,
```
Comment (unho) [inline pootle/apps/pootle_statistics/utils.py]: `include_anonymous` is more explanatory.

Comment (phlax) [inline pootle/apps/pootle_statistics/utils.py]: also more typing - i think anon is pretty explicit

File: tests/statistics/stats_utils.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    assert anon not in contribs.user_qs
+    assert (
+        sorted(contribs)
+        == sorted(contribs.contributors)
+        == sorted(
+            set(contribs.user_qs.filter(contribs.user_filters)
+                                .values_list("username", flat=True)))
+        == sorted(
+            set(contribs.user_qs.filter(submission__gt=0)
+                                .values_list("username", flat=True))))
```
Comment (unho) [inline tests/statistics/stats_utils.py]: Can't this be split into separate assert calls? Also perhaps the long queries can be stored in variables, just for ease reading.

File: pootle/apps/pootle_statistics/utils.py (hunk 1 above)
Comment (unho) [inline pootle/apps/pootle_statistics/utils.py]: I disagree.

File: pootle/apps/pootle_statistics/utils.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    def time_filters(self):
+        q = Q()
+        if self.since is not None:
+            q = q & Q(submission__creation_time__gte=self.since)
+        if self.until is not None:
+            q = q & Q(submission__creation_time__lte=self.until)
+        return q
+
+    @property
+    def query(self):
```
Comment (unho) [inline pootle/apps/pootle_statistics/utils.py]: Perhaps better `all_filters`?

File: pootle/apps/pootle_statistics/utils.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    def __init__(self, include_anon=False,
+                 project_codes=None, language_codes=None,
+                 since=None, until=None, sort_by=None):
+        self.include_anon = include_anon
+        self.project_codes = project_codes
+        self.language_codes = language_codes
+        self.since = since
+        self.until = until
+        self.sort_by = sort_by
```
Comment (unho) [inline pootle/apps/pootle_statistics/utils.py]: `sort_by` should default to `"name"`.

File: tests/statistics/stats_utils.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+from pootle_statistics.utils import Contributors
+from pootle_store.models import Store
+
+
+User = get_user_model()
+
+
+def _contributors_list(contribs,
+                       projects=None, languages=None,
+                       since=None, until=None):
```
Comment (unho) [inline tests/statistics/stats_utils.py]: No sort_by?

File: tests/statistics/stats_utils.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    subs = []
+    for user in contribs.user_qs.distinct():
+        q = Q(submitter=user)
+        if projects:
+            q = q & Q(
+                translation_project__project__code__in=projects)
```
Comment (unho) [inline tests/statistics/stats_utils.py]: Where is `submission__`?

Comment (unho) [inline tests/statistics/stats_utils.py]: nvm

File: tests/statistics/stats_utils.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        if languages:
+            q = q & Q(
+                translation_project__language__code__in=languages)
+        if since or until:
+            _q = Q()
```
Comment (unho) [inline tests/statistics/stats_utils.py]: Is it necessary to use `_q` when `q` is already available?

File: tests/statistics/stats_utils.py (hunk 2 above)
Comment (phlax) [inline tests/statistics/stats_utils.py]: they all equal the same thing - i think the test is more explicit this way

File: tests/statistics/stats_utils.py (hunk 5 above)
Comment (phlax) [inline tests/statistics/stats_utils.py]: its not tested yet

File: tests/statistics/stats_utils.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            if since:
+                _q = _q & Q(creation_time__gte=since)
+            if until:
+                _q = _q & Q(creation_time__lte=until)
+            q = q & _q
+        submissions = Submission.objects.filter(q)
```
Comment (unho) [inline tests/statistics/stats_utils.py]: I would call `.count()` here, just to call once.

File: tests/statistics/stats_utils.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        if submissions.count():
+            subs.append((user.username, submissions.count()))
+    return OrderedDict(sorted(subs, key=lambda i: (-i[1], i[0])))
```
Comment (unho) [inline tests/statistics/stats_utils.py]: We do not necessarily want to sort like this. If sorting by name it would be only `key=lambda i: i[0]`.

Comment (phlax) [inline tests/statistics/stats_utils.py]: again, sorting is not yet tested

File: tests/statistics/stats_utils.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+
+User = get_user_model()
+
+
+def _contributors_list(contribs,
+                       projects=None, languages=None,
+                       since=None, until=None):
+    subs = []
+    for user in contribs.user_qs.distinct():
```
Comment (unho) [inline tests/statistics/stats_utils.py]: `distinct()`?

File: tests/statistics/stats_utils.py (hunk 7 above)
Comment (phlax) [inline tests/statistics/stats_utils.py]: im not 100% sure if it necessary (i think it is - it seemed to be collapsing the query before - not sure) but it structures the query differently

File: tests/statistics/stats_utils.py (hunk 10 above)
Comment (phlax) [inline tests/statistics/stats_utils.py]: ??

Comment (unho) [inline tests/statistics/stats_utils.py]: All users are distinct already, you get the same with `.all()`.

File: tests/statistics/stats_utils.py (hunk 7 above)
Comment (unho) [inline tests/statistics/stats_utils.py]: Feels hacky. At least I would like to see some short comment explaining this.

File: tests/statistics/stats_utils.py (hunk 10 above)
Comment (phlax) [inline tests/statistics/stats_utils.py]: not always - altho to be fair i have now changed user_qs

try doing this

```
subs = User.objects.filter(submission__gt=0)
print subs.count()
print subs.distinct().count()
```

File: pootle/apps/pootle_statistics/utils.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+
+User = get_user_model()
+
+
+class Contributors(object):
+
+    def __init__(self, include_anon=False,
+                 project_codes=None, language_codes=None,
+                 since=None, until=None, sort_by="username"):
```
Comment (unho) [inline pootle/apps/pootle_statistics/utils.py]: I don't feel it is a good idea to put a mutable type here as default. Better stick to `None` and do some checking before assigning.

File: pootle/apps/pootle_statistics/utils.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        self.project_codes = project_codes
+        self.language_codes = language_codes
+        self.since = since
+        self.until = until
+        self.sort_by = sort_by
+
+    @property
+    def user_qs(self):
+        if self.include_anon:
+            return User.objects.exclude(username__in=["system", "default"])
```
Comment (unho) [inline pootle/apps/pootle_statistics/utils.py]: I have the feeling this should be moved to `UserManager`.

File: pootle/apps/pootle_statistics/utils.py (hunk 11 above)
Comment (phlax) [inline pootle/apps/pootle_statistics/utils.py]: ??

you told me to put it here - besides strings **immutable**

File: pootle/apps/pootle_statistics/utils.py (hunk 12 above)
Comment (phlax) [inline pootle/apps/pootle_statistics/utils.py]: no - this is relevant to this class - unless we have a need for this pattern elsewhere im going to keep it here for now

**Comment by unho** [comment]
gtm once tests for no contribs in range are added, and Travis goes green.

**Pull request merged**